peer/brontide.go

```diff
@@ -3136,19 +3136,18 @@ func (p *Brontide) tryLinkShutdown(cid lnwire.ChannelID) error {
 		return ErrChannelNotFound
 	}
 
-	// Else, the link exists, so attempt to trigger shutdown. If this
-	// fails, we'll send an error message to the remote peer.
-	if err := chanLink.ShutdownIfChannelClean(); err != nil {
-		return err
-	}
-
-	// Next, we remove the link from the switch to shut down all of the
-	// link's goroutines and remove it from the switch's internal maps. We
-	// don't call WipeChannel as the channel must still be in the
-	// activeChannels map to process coop close messages.
-	p.cfg.Switch.RemoveLink(cid)
-
-	return nil
+	return chanLink.Flush(func() {
```

Comment:
Some awareness/signal needs to be added to the `chancloser` or we'll run into spec violations. Consider the case where we're the responder in the coop close flow:

1. We receive a `Shutdown` and call `fetchActiveChanCloser`.
2. `fetchActiveChanCloser` calls `tryLinkShutdown` which calls `chanLink.Flush`. `Flush` is non-blocking so this returns immediately. We can't make this blocking as this would prevent other coop close requests from continuing.
3. `ProcessCloseMsg` is called on the received `Shutdown` message. This will make us send a `Shutdown` to the peer. 
4. It's possible that the `ChannelLink` sends an htlc after this in `handleDownstreamUpdateAdd`. This would be a spec violation.

The case where we are the coop close initiator leads to the same scenario above, but the initiator will actually send a premature `ClosingSigned` when the `chancloser` receives the peer's `Shutdown`.

`Shutdown` should only be sent if it's not possible for the link to send a htlc after `Shutdown` due to concurrency.

Reply:
Ok If I understand this correctly you're saying:

It's not sufficient to handle this in `EligibleToForward` because it may have already been added to the downstream mailbox but not processed by the htlcManager. It must be handled in the actual `handleDownstreamPkt` itself otherwise something may get into the link process queue, then we flush and send shutdown to our peer, then when we process that downstream add we still send it out because it was already in the queue.

I don't think that automatically implies we need to add something to the ChanCloser, but I do think it means that something needs to change here.

Reply:
> It's not sufficient to handle this in EligibleToForward because it may have already been added to the downstream mailbox but not processed by the htlcManager

My understanding is that the change in `EligibleToForward` is _necessary_ but not sufficient. We also need a way to cancel back anything that's un-ack'd that might be sitting in the mailbox. We can do that by doing a similar flush check in `handleDownstreamUpdateAdd`, which'll then allow us to cancel stuff within the mailbox batch to the switch.